internal/throttler/throttler.go

```diff
@@ -0,0 +1,111 @@
+//go:generate mockgen -source throttler.go -destination ../mocks/mock_throttler.go -package mocks
+
+package throttler
+
+import (
+	"context"
+	"time"
+
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+type Throttler interface {
+	Close()
+	Throttle(context.Context)
+}
+
+type NoopThrottler struct{}
+
+var _ Throttler = (*NoopThrottler)(nil)
+
+func (r *NoopThrottler) Throttle(ctx context.Context) {
+}
+
+func (r *NoopThrottler) Close() {
+}
+
+// throttler implements a throttling mechanism that can be used to control the rate of dispatched sub problems in FGA queries.
+// Throttling will start to kick in when the dispatch count exceeds the configured dispatch threshold.
+type throttler struct {
+	ticker          *time.Ticker
+	throttlingQueue chan struct{}
+	done            chan struct{}
+}
+
+// NewDispatchThrottler constructs a throttler which can be used to control the rate of dispatched sub problems in FGA queries.
+func NewDispatchThrottler(frequency time.Duration) Throttler {
+	return newDispatchThrottler(frequency)
+}
+
+// Returns a throttler instead of Throttler for testing purpose to be used internally.
+func newDispatchThrottler(frequency time.Duration) *throttler {
```

Comment:
why not incorporate this with NewDispatchThrottler function above?

Reply:
Because we want to use things like `nonBlockingSend` in test and also this method returns a internal implementation of `throttler`. The `NewDispatchThrottler` returns only the `Throttler` interface.